### ISSUE TITLE
Memory leak when publishing to RTMP server?

### ISSUE THREAD

**Issue by srajca** [bytedeco/javacv#1351]
Hello, we are having troubles with the FFmpegFrameRecorder, I think. The software uses all of memory and crashes when trying to stream to RTMP server. Is there a known leak or a solution to the problem? The memory grew to 4GB in matter of seconds at which point the software had to be terminated.

We are using the 1.5.2. version of JavaCV.
This is our class for streaming to RTMP:
```import org.bytedeco.ffmpeg.global.avcodec;
import org.bytedeco.ffmpeg.global.avutil;
import org.bytedeco.javacv.FFmpegFrameRecorder;
import org.bytedeco.javacv.Frame;
import org.bytedeco.javacv.FrameRecorder;

import java.util.Deque;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class VideoStreamer implements VideoOutput {

    private FrameRecorder mRecorder;
    private final String mUri;
    private final String mPreset;
    private final String mTune;
    private final int    mBitrate;

    private final Deque<Frame> mFrameBuffer;

    private boolean     mHasBeenInitialized = false;

    private ScheduledExecutorService mExecutor;

    private VideoStreamer(String outUri, String preset, String tune, int bitrate) {
        mUri = outUri;
        mPreset = preset;
        mTune = tune;
        mBitrate = bitrate;
        mFrameBuffer = new ConcurrentLinkedDeque<>();
    }

    private FrameRecorder createFrameRecorder(String outUrl, int width, int height, int videoBitrate, double framerate, int audioBitrate, int numAudioChannels) {

        FFmpegFrameRecorder rec = new FFmpegFrameRecorder(outUrl, width, height, numAudioChannels);
//        rec.setInterleaved(true);
        rec.setVideoCodecName("libx264");
        rec.setVideoCodec(avcodec.AV_CODEC_ID_H264);
        rec.setFormat("flv");
        rec.setVideoBitrate(mBitrate);//FIX THIS FOR EITHER A 1080p or 720p OR ADD OPTIONS IN SETTINGS
        rec.setImageWidth(width);
        rec.setImageHeight(height);

        // tradeoff between quality and encode speed
        // possible values are ultrafast,superfast, veryfast, faster, fast,
        // medium, slow, slower, veryslow
        // ultrafast offers us the least amount of compression (lower encoder
        // CPU) at the cost of a larger stream size
        // at the other end, veryslow provides the best compression (high
        // encoder CPU) while lowering the stream size
        // (see: https://trac.ffmpeg.org/wiki/Encode/H.264)
        rec.setVideoOption("preset", mPreset);

        // decrease "startup" latency in FFMPEG (see: https://trac.ffmpeg.org/wiki/StreamingGuide )
        rec.setVideoOption("tune", mTune);

        // Constant Rate Factor (see: https://trac.ffmpeg.org/wiki/Encode/H.264)
//        rec.setVideoOption("crf", "28");

        rec.setFrameRate(framerate);
        rec.setGopSize((int) framerate * 2);
//        rec.setVideoOption("keyint_min", String.valueOf((int) framerate * 2));
//        rec.setVideoOption("sc_threshold", "0");
        rec.setPixelFormat(avutil.AV_PIX_FMT_YUV420P);
        rec.setAudioCodecName("aac");
        rec.setAudioBitrate(120000); // forced 120kb/s
        rec.setAudioChannels(2); // Changed to 2 channel for youtube
//        rec.setAudioChannels(numAudioChannels); // But we can use more, at least as the input video

        return rec;
    }


    @Override
    public void init(FluxFormat format) throws CannotInitOutputException {

        mRecorder = createFrameRecorder(mUri,
                format.imageWidth,
                format.imageHeight,
                format.videoBitrate,
                format.videoFramerate,
                format.audioBitrate,
                format.audioFormat.getChannels());

        try {
            mRecorder.start();
        } catch (FrameRecorder.Exception e) {
            throw new CannotInitOutputException("Failed to initialize the stream with url: \""+mUri+"\".",e);
        }

        mHasBeenInitialized = true;

        mExecutor = new ScheduledSingleThreadExecutorWithException(new AdjustableThreadFactory("VideoStreamer - stream"));

    }

    int test = 0;

    @Override
    public void feed(Frame f) {
        if (mHasBeenInitialized)
            mExecutor.submit(() -> stream(f));
    }

    // Don't need it for the moment
    @Override
    public void start() {
        if (!mHasBeenInitialized)
            throw new NotInitializedException("Initialization failed or never been called");
    }

    @Override
    public void stop() {
        mExecutor.shutdownNow();
        try {
            mExecutor.awaitTermination(1, TimeUnit.SECONDS);
            mRecorder.stop();
            mRecorder.release();
        } catch (FrameRecorder.Exception e) {
            throw new MediaException("Cannot close the video streamer.", e);
        } catch (InterruptedException e) {
            //e.printStackTrace(); // not an error, if stoped is pressed
        }
    }

    @Override
    public void pause() {
        if (!mHasBeenInitialized)
            throw new NotInitializedException("Initialization failed or never been called");
    }

    @Override
    public void resume() {
        if (!mHasBeenInitialized)
            throw new NotInitializedException("Initialization failed or never been called");
    }

    @Override
    public void elapsedTimeChanged(long millisSecondFromBeginning) {
        mFrameBuffer.clear();
    }

    // This method is called by the executor service, it's needed to push frame to the recorder in another thread.
    // If we do it in the feed() method, it's slowing down the whole pipeline, who needs to be the most responsive
    // as possible to avoid lag (specially in the preview)
    public void stream(Frame f) {
//        Frame f = mFrameBuffer.poll();
        if (f == null)
            return;
        try {
//            if (mRecorder.getAudioChannels() == 0
            mRecorder.record(f);
        } catch (FrameRecorder.Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
    }

    public static class Builder {
        private String uri;
        private String preset;
        private String tune;
        private int    bitrate;

        public VideoStreamer build() {
            return new VideoStreamer(uri, preset, tune, bitrate);
        }

        public Builder setBitrate(int bitrate) {
            this.bitrate = bitrate;
            return this;
        }

        public Builder setPreset(String preset) {
            this.preset = preset;
            return this;
        }

        public Builder setTune(String tune) {
            if (tune.compareTo("none") == 0) // Handle none parameter, shadow it with an empty string.
                tune = "";
            this.tune = tune;
            return this;
        }

        public Builder setUri(String uri) {
            this.uri = uri;
            return this;
        }
    }
}
```

Any thoughts on the matter? Thank you

**Comment by saudet** [comment]
`FFmpegFrameRecorder` is not thread-safe. It's possible that `stop()` doesn't get called at the right moment and that could leak memory. Try to call `stop()` in the same thread that calls `start()`.

Apart from that, please provide data about memory usage that your profiler gives you. It's unlikely that anyone will be able to help you without any additional information.

**Comment by davepaiva** [comment]
Maybe it is a simiar issue?
#1341

**Comment by vijayakumarbathini** [comment]
Can you please share heap dump when memory has grown up. So Its easy to identify which thread is eating the most of memory.

Possible way to collect heap dump 

jmap -dump:file=heapdump.hprof,format=b <pid>

**Comment by srajca** [comment]
The problem we had actually originated from

public void feed(Frame f)  and   public void stream() . We added this in hopes of bettering the pipeline but ended up clogging the memory. This is where the leak was formed, it had nothing to do with FFmpegFrameRecorder.
There  was a call of mExecutor.submit(() -> stream(f)) before but such a solution caused  OutOfMemoryError if FFmpegFrameRecorder didn’t keep up with sending frames in the stream(f) method.
Calls of () -> stream(f) accumulated in the executor service line and slowed the whole app down and caused a memory leak. frameQueue acts as a fixed-length buffer which prevents several thousand frames from hanging in the memory.

Instead of getting a frame as a parameter we get it from the frameQueue now.
        Frame f = frameQueue.poll();
        if (f == null)
            return;

Thank you all.
